python/paddle/_typing/shape.py

```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Tuple, Union
+
+from typing_extensions import TypeAlias
+
+from .. import Tensor
+
+DynamicShapeLike: TypeAlias = Union[
+    Tuple[Optional[Union[int, Tensor]], ...],
+    list[Optional[Union[int, Tensor]]],
```

Comment:
```suggestion
    Tuple[Union[int, Tensor, None], ...],
    list[Union[int, Tensor, None]],
```

这样明显语义更清晰

另外 23 行的 `list[xxx]` 在 Python 3.8 不会有问题吗？Python 3.9 才有的 builtin list 可作为类型提示

是因为没有任何地方 import 吧，所有代码都没测到，加一个单测保证这些 type 运行时没问题吧，现在应该运行时就会挂

Reply:
这里手误写错了，不过我在本地测试过，这些代码都没问题，3.8.0可以使用list作为类型提示，现在添加了mypy进行测试